### ISSUE TITLE
FPU support?

### ISSUE THREAD

**Issue by davidgiven** [kstenerud/Musashi#30]
I notice that the version of Musashi in mame has support for more architectures, and the FPU; are there any plans to import that version here? (I'm using the emulator to run compiler tests, and some of them use floating point.)

**Comment by zelurker** [comment]
I have made a fork of this repository where I tried to import a few patches from mame if you are interested, they include the support for the fpu, but only the beginning, I mean my goal when merging the patch was keeping musashi outside of mame internals but it becomes hard to do very quickly, around version 125 only.
It's a basic support, no definition for a specific cpu+fpu combination, but the fpu opcodes are handled and it compiles so it should be usable for what you have. I also updated the example to be able to compile it in linux so that I could quickly test that everything at least compiles.

**Comment by bebbo** [comment]
the pc is not yet updated correctly, the additional read words are not added to the pc.

**Comment by zelurker** [comment]
Would you have some example code using the 040 fpu ? I looked on the wab,
but it seems it's not easy to find.
I mean something usable without a 68040 assembler preferably, instructions
+ opcodes if possible...
Just for info, I never see this kind of code...

As previously said there are still a lot of patches to merge, but the big
problem is that after this point musashi was made very dependant on the
mame core and it becomes very difficult to follow the patches after that if
you want to remain independent from mame, it's not impossible, but very
hard.

**Comment by bebbo** [comment]
> Would you have some example code using the 040 fpu ? I looked on the wab, but it seems it's not easy to find. I mean something usable without a 68040 assembler preferably, instructions + opcodes if possible... Just for info, I never see this kind of code...

do you need https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=2ahUKEwj0g-mH4qbmAhXLsaQKHUkCDjcQFjAAegQIAxAC&url=https%3A%2F%2Fwww.nxp.com%2Ffiles-static%2Farchives%2Fdoc%2Fref_manual%2FM68000PRM.pdf&usg=AOvVaw3HdjcVYsrGb6tkIzqytdzC or https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=2ahUKEwiF0oij4qbmAhXKjKQKHW83DlEQFjAAegQIAhAC&url=http%3A%2F%2Fbitsavers.trailing-edge.com%2Fcomponents%2Fmotorola%2F68000%2FMC68881_MC68882_Floating-Point_Coprocessor_Users_Manual_1ed_1987.pdf&usg=AOvVaw1peNHzJ7TEGbmW9qB-Wy3t ?


> As previously said there are still a lot of patches to merge, ...

I also implemented some more instructions - so it runs what I want it to run. I'm using https://github.com/bebbo/amitools so I integrated you upated version into it. To create and run programs I'm using the Amiga gcc toolchain https://github.com/bebbo/amiga-gcc

**Comment by zelurker** [comment]
Yeah I know m68000prm but I had hoped for some short code examples, easy to
read a to follow, oh well forget it...
Well let us know if you merge some more patches then Karl always welcomes
more pull requests, that's the way to make them public !
On my side my main interest was for 68000 to 68020, so musashi as it is now
is enough for me, but I might take another look anyway at merging more
patches later...

**Comment by bebbo** [comment]
an convenient way to "get some fpu asm code" is to use https://franke.ms/cex/
enter some c /c++ code and use -m68040 as compiler options and you'll see the corresponding asm.

**Comment by zelurker** [comment]
Interesting site, thanks, but it doesn't seem to work for this purpose, if
you type
double x = 1.5*num;
you get 2 function calls from a library instead of the direct instructions :
move.l (8,a5),-(sp)
jbsr ___extendsfdf2
addq.l #4,sp
clr.l -(sp)
move.l #1073217536,-(sp)
move.l d1,-(sp)
move.l d0,-(sp)
jbsr ___muldf3
lea (16,sp),sp
move.l d0,(-8,a5)
move.l d1,(-4,a5)

But I'll keep the url of the site, thanks !

**Comment by michalsc** [comment]
You need to pass compiler options (e.g. -m68040 -m68881), otherwise 68000 mode without FPU will be used

**Comment by bebbo** [comment]
try this: https://franke.ms/cex/z/m0HstT

**Comment by zelurker** [comment]
Nice but without the opcodes, it's useless.
I tried "compile the binary and disassemble the output", but it says "no
output, objdump returned 1", I guess it doesn't like the -m68040 option.
By the way I had used the -m68040 option too of course, no idea why it
produces the instructions here and the library calls instead before.
Anyway... !

**Comment by michalsc** [comment]
> Would you have some example code using the 040 fpu ? I looked on the wab, but it seems it's not easy to find. I mean something usable without a 68040 assembler preferably, instructions + opcodes if possible... Just for info, I never see this kind of code... 

If you like you can have a look at my M68k JIT for ARM processors, especially the examples added there: https://github.com/michalsc/Emu68/tree/master/Examples. Two of the examples, Buddha and SmallPT, use FPU very intensively. If you like I can provide you disassemble output of both of them - asm code plus opcodes.

**Comment by bebbo** [comment]
> 
> Nice but without the opcodes, it's useless. I tried "compile the binary and disassemble the output", but it says "no output, objdump returned 1", I guess it doesn't like the -m68040 option. By the way I had used the -m68040 option too of course, no idea why it produces the instructions here and the library calls instead before. Anyway... !

try again - I updated the config

**Comment by zelurker** [comment]
Le mer. 11 déc. 2019 à 10:58, bebbo <notifications@github.com> a écrit :

> try again - I updated the config


Thanks for the objdump fix, now it's perfect !
Well to come back to what started this topic, all the instructions I tried
so far increase correctly the pc.
I didn't try all of them of course, just a copy of this simple compilation.

I tried this to have some operand after the instruction :

f22d 5444 0008 which translates to fdmoved 8(a5),fp0
works fine, it reads correctly the 8 offset using read_immediate which
correctly increases the pc.

Anyway might be a good idea to try to add some example for the fpu too,
I'll look more into this later...

**Comment by bebbo** [comment]
> Well to come back to what started this topic, all the instructions I tried so far increase correctly the pc. I didn't try all of them of course, just a copy of this simple compilation. I tried this to have some operand after the instruction : f22d 5444 0008 which translates to fdmoved 8(a5),fp0 works fine, it reads correctly the 8 offset using read_immediate which correctly increases the pc. Anyway might be a good idea to try to add some example for the fpu too, I'll look more into this later...

I'm using vamos/musashi to run the gcc-check tests (> 14000) and for now there are no open issues related to musashi.

**Comment by zelurker** [comment]
At least it gave me the motivation to propose a new pull request to Karl,
it's about some more patches from mame, until the emulation of the 68030
this time. It's just the beginning, for the 030, but it's better than
nothing.
The patches are harder to process, but it's not so bad... !
Plenty of new stuff to test then...
Sorry no update for the fpu for now...

**Comment by jotd666** [comment]
I've modded the FPU source code, supporting more 68040 FPU instructions.

- fsdivs / fsmuls (basically same code as fdiv/fmul since source is already converted as floatx80)
- fixed error message (fprintf should be vfprintf for varargs)
- added more modes for fmovem: ex: fmovem fp0/fp1,(-20,a6) and the read counterpart. Tricky as it requires modification of WRITE_EA_FPE and READ_EA_FPE interfaces (cannot just pass ea)
- added fscc (set 8 bits of register if condition is met else clears them)
- ftrap: added +2 to PC, but didn't implement the trap. At least if no exception occurs the code doesn't stumble on illegal instruction.

Someone is interested?

**Comment by bebbo** [comment]
@jotd: what's your repo?

**Comment by davidgiven** [comment]
If you mean interested _in general_, yes --- I'm using Musashi to run compiler tests for the Amsterdam Compiler Kit (at http://github.com/davidgiven/ack), and having FPU support would allow testing floating point code.

**Comment by jotd666** [comment]
@bebbo I don't have any repo right now. I'm willing to contribute but I'm still unsure about 100% correctness of my additions. My app isn't running properly yet (it may be because of other incompatibilities)

**Comment by jotd666** [comment]
ok, update after working all week on that, with success. I updated FPU but also some exotic scalar instructions like CHK2/CMP2

m68k_in.c:

- fixed all CHK2/CMP2 instructions to support signed bounds. Previously it only worked with unsigned. Also sped up a bit
  by changing | by || in Z evaluation formula
- cptrapcc opcodes are still unsupported, but if no exception occurs, at least PC is properly updated
- trapt/trapcc 16 & 32 instructions: PC is updated before exception is triggered (or if exception isn't triggered). This fixes the
  return address value (just in case code returns from trap with RTE) and also the stackframe (XDAda 68040 compiler
  read the data parameter to check exception type, wrong exception type is read if stackframe is incorrect)
  
m68kfpu.c:

- fatalerror message fprintf => vfprintf
- added address register + displacement mode (mode 5, READ_EA_FPE/WRITE_EA_FPE), also changed prototype to it can be called with this mode 
  from FMOVEM loop
- added immediate mode (READ_EA_FPE)
- fpgen_rm_reg (FMOVE to register): added a lot of missing condition code update (SET_CONDITION_CODES(REG_FP[dst]))
- added FSDIVS and FSMULS (easy as it's the same case as FDIV and FMUL). Updated condition codes for FDIV
- fmove_fpcr: connected softfloat 4 rounding modes to this instruction, that didn't set rounding at all, when it's
  perfectly emulated in softfloat module (just added:  float_rounding_mode = (REG_FPCR >> 4) & 0x3;)
- fmovem: added address register+displacement mode (READ_EA_FPE/WRITE_EA_FPE interface had to be changed)
- added FScc instruction

The number of cycles isn't accurate. But the modified sources allowed to run a complex avionic program that computes
a plane trajectory, exactly the same way as it's done using Basilisk UAE emulation core (and even better since Basilisk
doesn't support proper rounding in FPU instructions)

**Comment by jotd666** [comment]
@bebbo I have created a new repo (completely copied from musashi) here: https://github.com/jotd666/musashi040

I suck at git, couldn't do something like a fork. I hope my changes will be officialized, so I can delete my repo afterwards

**Comment by bebbo** [comment]
@jotd666 thanks a lot - I cloned your repo and will look into it

**Comment by kstenerud** [comment]
Awesome! I look forward to it!

**Comment by jotd666** [comment]
I hope I didn't create too much noise with my changes. As I forked main
repo, created another one on my account, and posted attached fixed files
too... I don't mind creating issues, with attached files, and let you
commit changes (or not).

Le lun. 9 mars 2020 à 08:29, Karl Stenerud <notifications@github.com> a
écrit :

**Comment by kstenerud** [comment]
I'm usually massively swamped, so every bit of assistance helps. I'm hoping @bebbo can verify from his test set, and then I'll add in the modifications.

But for future, you'd probably save a lot of work on yourself by just forking the repo, making changes on your copy of the repo, and then opening a PR from there to here. Github makes it pretty painless :)

**Comment by jotd666** [comment]
Yes, I'm still getting used to github/git. Long time clearcase/UCM user...

Needless to say your emulator is saving my life at work right now because
of the nice licence :)

Le mar. 10 mars 2020 à 22:35, Karl Stenerud <notifications@github.com> a
écrit :

**Comment by shoggoth77** [comment]
It seems to me that there's no clean way to turn off FPU support, and it adds quite a lot of code. That's sort of a regression for my particular case. It would be nice if this could be optional.
